add quantized GELU activation

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#34396 add quantized GELU activation**

Summary:

Adds quantized implementation of GELU. Since the floating point implementation
doesn't currently have the inplace and out flavors, we also skip them here for
now.

Would love reviewers' feedback on if we need to do anything smarter with
scale+zp here, especially around -2 < x < 0.  Will address that either
as followup on this PR or as a separate PR, depending of what people
say and whether this is important.

Test Plan:

```
python test/test_quantized.py
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20371751](https://our.internmc.facebook.com/intern/diff/D20371751)

vkuzo committed:
add quantized GELU activation

Summary:

Adds quantized implementation of GELU. Since the floating point implementation
doesn't currently have the inplace and out flavors, we also skip them here for
now.

Would love reviewers' feedback on if we need to do anything smarter with
scale+zp here, especially around -2 < x < 0.  Will address that either
as followup on this PR or as a separate PR, depending of what people
say and whether this is important.

Test Plan:

```
python test/test_quantized.py
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "add quantized GELU activation"


Summary:

Adds quantized implementation of GELU. Since the floating point implementation
doesn't currently have the inplace and out flavors, we also skip them here for
now.

Would love reviewers' feedback on if we need to do anything smarter with
scale+zp here, especially around -2 < x < 0.  Will address that either
as followup on this PR or as a separate PR, depending of what people
say and whether this is important.

Test Plan:

```
python test/test_quantized.py
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

z-a-f commented on test/test_quantized.py:
```diff
@@ -169,6 +170,31 @@ def test_qrelu_leaky(self, X, alpha):
+        # calculate GT
+        dqX = qX.dequantize()
+        # GELU
+        dqY_hat = dqX * 0.5 * (1.0 + (dqX / math.sqrt(2)).erf())
```
Isn't it better to use the floating point version of the gelu here (`nn.functional.gelu`).

z-a-f reviewed (approved):
LGTM + nit

vkuzo committed:
Update on "add quantized GELU activation"


Summary:

Adds quantized implementation of GELU. Since the floating point implementation
doesn't currently have the inplace and out flavors, we also skip them here for
now.

Would love reviewers' feedback on if we need to do anything smarter with
scale+zp here, especially around -2 < x < 0.  Will address that either
as followup on this PR or as a separate PR, depending of what people
say and whether this is important.

Test Plan:

```
python test/test_quantized.py
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20371751](https://our.internmc.facebook.com/intern/diff/D20371751)

[ghstack-poisoned]

supriyar commented on test/test_quantized.py:
```diff
@@ -169,6 +170,31 @@ def test_qrelu_leaky(self, X, alpha):
+        qX = torch.quantize_per_tensor(X, scale=scale, zero_point=zero_point,
+                                       dtype=torch_type)
+
+        # calculate GT
```
nit: Not sure what you mean by GT here

vkuzo committed:
Update on "add quantized GELU activation"


Summary:

Adds quantized implementation of GELU. Since the floating point implementation
doesn't currently have the inplace and out flavors, we also skip them here for
now.

Would love reviewers' feedback on if we need to do anything smarter with
scale+zp here, especially around -2 < x < 0.  Will address that either
as followup on this PR or as a separate PR, depending of what people
say and whether this is important.

Test Plan:

```
python test/test_quantized.py
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20371751](https://our.internmc.facebook.com/intern/diff/D20371751)

[ghstack-poisoned]

vkuzo committed:
Update on "add quantized GELU activation"


Summary:

Adds quantized implementation of GELU. Since the floating point implementation
doesn't currently have the inplace and out flavors, we also skip them here for
now.

Would love reviewers' feedback on if we need to do anything smarter with
scale+zp here, especially around -2 < x < 0.  Will address that either
as followup on this PR or as a separate PR, depending of what people
say and whether this is important.

Test Plan:

```
python test/test_quantized.py
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20371751](https://our.internmc.facebook.com/intern/diff/D20371751)

[ghstack-poisoned]

vkuzo committed:
Update on "add quantized GELU activation"


Summary:

Adds quantized implementation of GELU. Since the floating point implementation
doesn't currently have the inplace and out flavors, we also skip them here for
now.

Would love reviewers' feedback on if we need to do anything smarter with
scale+zp here, especially around -2 < x < 0.  Will address that either
as followup on this PR or as a separate PR, depending of what people
say and whether this is important.

Test Plan:

```
python test/test_quantized.py
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20371751](https://our.internmc.facebook.com/intern/diff/D20371751)

[ghstack-poisoned]

vkuzo committed:
Update on "add quantized GELU activation"


Summary:

Adds quantized implementation of GELU. Since the floating point implementation
doesn't currently have the inplace and out flavors, we also skip them here for
now.

Would love reviewers' feedback on if we need to do anything smarter with
scale+zp here, especially around -2 < x < 0.  Will address that either
as followup on this PR or as a separate PR, depending of what people
say and whether this is important.

Test Plan:

```
python test/test_quantized.py
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20371751](https://our.internmc.facebook.com/intern/diff/D20371751)

[ghstack-poisoned]

vkuzo committed:
Update on "add quantized GELU activation"


Summary:

Adds quantized implementation of GELU. Since the floating point implementation
doesn't currently have the inplace and out flavors, we also skip them here for
now.

Would love reviewers' feedback on if we need to do anything smarter with
scale+zp here, especially around -2 < x < 0.  Will address that either
as followup on this PR or as a separate PR, depending of what people
say and whether this is important.

Test Plan:

```
python test/test_quantized.py
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20371751](https://our.internmc.facebook.com/intern/diff/D20371751)

[ghstack-poisoned]

vkuzo committed:
Update on "add quantized GELU activation"


Summary:

Adds quantized implementation of GELU. Since the floating point implementation
doesn't currently have the inplace and out flavors, we also skip them here for
now.

Would love reviewers' feedback on if we need to do anything smarter with
scale+zp here, especially around -2 < x < 0.  Will address that either
as followup on this PR or as a separate PR, depending of what people
say and whether this is important.

Test Plan:

```
python test/test_quantized.py
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20371751](https://our.internmc.facebook.com/intern/diff/D20371751)

[ghstack-poisoned]

raghuramank100 commented:
Can we also add benchmark results showing the performance improvement of the quantized impl over float?

vkuzo commented:
> Can we also add benchmark results showing the performance improvement of the quantized impl over float?

will do, need to chase down the real test failure first as seems like the math is wrong in some CircleCI environments

vkuzo committed:
Update on "add quantized GELU activation"


Summary:

Adds quantized implementation of GELU. Since the floating point implementation
doesn't currently have the inplace and out flavors, we also skip them here for
now.

Would love reviewers' feedback on if we need to do anything smarter with
scale+zp here, especially around -2 < x < 0.  Will address that either
as followup on this PR or as a separate PR, depending of what people
say and whether this is important.

Test Plan:

```
python test/test_quantized.py
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20371751](https://our.internmc.facebook.com/intern/diff/D20371751)

[ghstack-poisoned]

vkuzo commented:
There are some flaky test failures here which need to be fixed before this can land.  My best guess is that `std::erf` is misbehaving on some platforms.  Closing for now since there is no immediate ask for this, but we will revisit when needed.
